languages/python/sqlalchemy-oso/tests/test_roles.py

```diff
@@ -312,88 +320,239 @@ def test_add_user_role(test_db_session):
+    read_repo_role = (
+        test_db_session.query(RepositoryRole)
+        .filter_by(user=john, repository=abbey_road)
+        .first()
+    )
+    org_owner_role = (
+        test_db_session.query(OrganizationRole)
+        .filter_by(user=john, organization=beatles)
+        .first()
+    )
+
+    # test base `resource_role_applies_to`
+    results = list(
+        oso.query_rule(
+            "resource_role_applies_to", abbey_road, Variable("role_resource")
+        )
+    )
+    assert len(results) == 1
+    assert results[0].get("bindings").get("role_resource") == abbey_road
+
+    # test custom `resource_role_applies_to` rules (for nested resources)
+    resource_role_applies_to_str = """resource_role_applies_to(repo: Repository, parent_org) if
+        parent_org := repo.organization and
+        parent_org matches Organization;
+        """
+    oso.load_str(resource_role_applies_to_str)
+    results = list(
+        oso.query_rule(
+            "resource_role_applies_to", abbey_road, Variable("role_resource")
+        )
+    )
+    results.sort(key=lambda x: x.get("bindings").get("role_resource").name)
+    assert len(results) == 2
+    assert results[0].get("bindings").get("role_resource") == abbey_road
+    assert results[1].get("bindings").get("role_resource") == beatles
+
+    # test `user_in_role` for RepositoryRole
+    results = list(oso.query_rule("user_in_role", john, Variable("role"), abbey_road))
+    assert len(results) == 1
+    assert results[0].get("bindings").get("role").name == "READ"
+
+    # test `user_in_role` for OrganizationRole
+    results = list(oso.query_rule("user_in_role", john, Variable("role"), beatles))
+    assert len(results) == 1
+    assert results[0].get("bindings").get("role").name == "OWNER"
+
+    # test `inherits_role` and `resource_role_order`
+    # make sure `inherits_role` returns nothing without a role order rule
+    results = list(
+        oso.query_rule("inherits_role", org_owner_role, Variable("inherited_role"))
+    )
+    assert len(results) == 0
+
+    # test role_order rule
+    role_order_str = 'organization_role_order(["OWNER", "MEMBER", "BILLING"]);'
+    oso.load_str(role_order_str)
+
+    results = list(
+        oso.query_rule("inherits_role", org_owner_role, Variable("inherited_role"))
+    )
+    results.sort(key=lambda x: x.get("bindings").get("inherited_role").name)
+    assert len(results) == 2
+    assert results[0].get("bindings").get("inherited_role").name == "BILLING"
+    assert results[1].get("bindings").get("inherited_role").name == "MEMBER"
+
+    # TODO: test `role_allow`
+    # make sure this query fails before any rules are added
+    results = list(oso.query_rule("role_allow", john, "READ", abbey_road))
+    assert len(results) == 0
+
+    # test basic `role_allow` rule
+    role_allow_str = """role_allow(role: RepositoryRole{name: "READ"}, "READ", repo: Repository) if
+                            role.repository.id = repo.id;"""
```

Comment:
To confirm: is this condition double-checking? We _should_ already know the role applies to the repo, right?

Reply:
Correct. I think I'm going to remove these.